File: src/main/java/algorithms/tree/103/no_103.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque
from idlelib.tree import TreeNode
from typing import List, Optional


class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        ans = []
        if root is None: return ans
        q = deque()
        q.append(root)
        needReverse = False
        while len(q) > 0:
            cnt = len(q)
            cur_level = []
            while cnt > 0:
                cur_node = q.pop()
                cur_level.append(cur_node.val)
                if cur_node.left is not None: q.appendleft(cur_node.left)
                if cur_node.right is not None: q.appendleft(cur_node.right)
                cnt -= 1
            if (needReverse):
                cur_level.reverse()
            ans.append(cur_level)
            needReverse = not needReverse
        return ans

File: src/main/java/algorithms/tree/no103/No103.java
package algorithms.tree;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;
import java.util.Queue;

/**
 * 103. 二叉树的锯齿形层序遍历
 * @author Sean Yu
 */
public class No103 {
}

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

/**
 * 思路：
 * 层序遍历的基础上，加上一个方向判断，如果是右，就将当前层序reverse加入结果集即可
 */
class Solution103 {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> ans = new ArrayList();
        if(root == null){
            return ans;
        }
        Queue<TreeNode> q = new ArrayDeque();
        boolean left = true;
        q.add(root);
        while(!q.isEmpty()){
            int cnt = q.size();
            List<Integer> list = new ArrayList();
            while(cnt > 0){
                cnt--;
                TreeNode cur = q.poll();
                list.add(cur.val);
                if(cur.left != null){
                    q.add(cur.left);
                }
                if(cur.right != null){
                    q.add(cur.right);
                }
            }
            if(!left){
                reverseList(list);
            }
            ans.add(list);
            left = !left;
        }
        return ans;
    }

    private void reverseList(List<Integer> list){
        int l = 0 , r = list.size()  - 1;
        while(l < r){
            int temp = list.get(r);
            list.set(r, list.get(l));
            list.set(l, temp);
            l++;
            r--;
        }
    }
}
